### ISSUE TITLE
Napi::TypeError::New is not an instanceof TypeError

### ISSUE THREAD

**Issue by rynz** [nodejs/node-addon-api#743]
Consider:
```
Napi::Value Rwar(const Napi::CallbackInfo& info) {
  Napi::Env env = info.Env();
  Napi::TypeError::New(env, "Wrong number of arguments")
      .ThrowAsJavaScriptException();
  return env.Null();
}
```
```
test('throws wrong number of arguments', () => {
  expect.assertions(1)
  try {
    addon.rwar()
  } catch (err) {
    expect(err).toBeInstanceOf(TypeError)
  }
})
```
For whatever reason I am getting:
```
● throws wrong number of arguments

    expect(received).toBeInstanceOf(expected)

    Expected constructor: TypeError
    Received constructor: TypeError

       6 |     addon.rwar()
       7 |   } catch (err) {
    >  8 |     expect(err).toBeInstanceOf(TypeError)
         |                 ^
       9 |   }
      10 | })
      11 | 
```
Whenever I check `err instanceof Error` or `err instanceof TypeError` both return `false`. Is there something I am doing wrong, or is this a bug?

**Comment by mhdawson** [comment]
I don't see anything obviously wrong with the node-addon-api or N-API code in core, looks like they should create a TypeError. What kind of object are you receiving and any more info in it that might help figure out what is going on?

**Comment by rynz** [comment]
@mhdawson I've created an example [rynz/node-addon-api-exception-example ](https://github.com/rynz/node-addon-api-exception-example)
 
That's the thing, it is receiving an Error/TypeError and the prototype chain looks okay. I've added a test to compare N-API vs Javascript debug.

```
  node-addon-api
    ✕ throws wrong number of arguments (2 ms)
  javascript
    ✓ throws wrong number of arguments (1 ms)

  ● node-addon-api › throws wrong number of arguments

    expect(received).toBeInstanceOf(expected)

    Expected constructor: TypeError
    Received constructor: TypeError

       8 |     } catch (err) {
       9 |       debugger
    > 10 |       expect(err).toBeInstanceOf(TypeError)
         |                   ^
      11 |     }
      12 |   })
      13 | })
```

**Comment by mhdawson** [comment]
Thanks for putting together the example, I'll try to take a look next week.

**Comment by NickNaso** [comment]
Hi everybody,
today I did some attempts using the example prepared by @rynz. 
I created a simple `index.js` file like reported below:

```js
'use strict'

const addon = require('./lib/rwar')
const assert = require('assert')

try {
  addon.rwar()
} catch (err) {
  assert.strictEqual(typeof TypeError == 'function', true)
  assert.strictEqual(err instanceof TypeError, true)
  assert.strictEqual(err instanceof Error, true)
  assert.strictEqual(typeof err.constructor === 'function' && err.constructor == TypeError, true)
  assert.strictEqual(typeof err.constructor === 'function' && err.constructor == Error, false)
}
```

I tried to replicate the checks that `jest` do on **[toBeInstanceOf](https://github.com/facebook/jest/blob/2a92e7f49fa35b219e5099d56b0179bccc1bf53e/packages/expect/src/matchers.ts#L280)** and they produced the expected results.

**Comment by mhdawson** [comment]
I was also just looking at this.  Using plain old instanceof gives the expected answer:
```JavaScript
const addon = require('../rwar')

try {
  addon.rwar()
} catch (err) {
  console.log(err);
  console.log(err instanceof TypeError);
}
```
-sh-4.2$ node test.js
TypeError: Wrong number of arguments
    at Object.<anonymous> (/home/mhdawson/test/node-addon-api-exception-example/lib/__tests__/test.js:4:9)
    at Module._compile (internal/modules/cjs/loader.js:1217:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1238:10)
    at Module.load (internal/modules/cjs/loader.js:1066:32)
    at Function.Module._load (internal/modules/cjs/loader.js:954:14)
    at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:71:12)
    at internal/main/run_main_module.js:17:47
true

```shell

```

**Comment by NickNaso** [comment]
I'm trying to execute the some tests with other test suite like:

- **node-tap**:
```js
'use strict'

const tap = require('tap')
const addon = require('./lib/rwar')

const test = tap.test

test('test 1', (t) => {
  try {
    addon.rwar()
  } catch (err) {
    t.match(err, TypeError)
  }
  t.done()
})

test('test 2', (t) => {
  const fn = function() {
    return addon.rwar()
  }
  t.throws(fn, TypeError)
  t.done()
})
```

- **jasmine**

```js
'use strict'

const addon = require('./lib/rwar')
describe('Test addon', function () {
  it('Schould be TypeError', function() {
    try {
      addon.hello()
    } catch (err) {
      expect(err).toBeInstanceOf(Error)
      expect(err).toBeInstanceOf(TypeError)
    }
  })
})
```
All continue to work as expected. Do you know if with **jest** we need some particular configuration?

**Comment by mhdawson** [comment]
I'm wondering if jest does some instrumentation of TypeError that's getting in the way.

**Comment by NickNaso** [comment]
Hi everybody,
I provided to open an issue on `jest` repo hope that hey give us some suggestion about. The issue is https://github.com/facebook/jest/issues/10189

**Comment by mhdawson** [comment]
@NickNaso great idea, I think we need somebody with knowledge of jest to comment.

**Comment by mhdawson** [comment]
The root cause may be related to the same issue in https://github.com/nodejs/node-addon-api/issues/764. The effects of JEST using different contexts.

**Comment by mhdawson** [comment]
Does seems related to JEST using contexts.  This test recreates the behavior reported:

```JavaScript
const addon = require('../rwar')
const vm = require('vm');

try {
  addon.rwar();
} catch (err) {
  console.log('Regular context:' +  (err instanceof TypeError));
}

try {
  const contextObject = { };
  vm.runInNewContext("addon.rwar()", contextObject);
} catch (err) {
  console.log('New context:' + (err instanceof TypeError));
}
```

**Comment by mhdawson** [comment]
@rynz you can read a bit more on the details in  #764 but this seems to be an expected result of Node.js  (not specific to N-API) not supporting multi-Context addons and the use of vm.runInNewContext by Jest.

**Comment by rynz** [comment]
@mhdawson thank you.

**Comment by mhdawson** [comment]
In terms of "multi-context" I think "context" might be a bit overloaded. I do know that there is/was work to let add-ons work nicely with Worker Threads such that you have ways to easily store data which is not global but instead specific to the worker thread from which the add-on is called. Sometimes I think I've seen this referred to as multi-context or something similar as well. That's one use case.

In terms of the question you raised in this issue, the use case is doing things like calling in  vm.runInNewContext and getting what you expect if it had been JavaScript code versus native code. 

Not sure this helped to clarify.  @gabrielschulhof do you have a better explanation?

**Comment by mhdawson** [comment]
Some discussion in the N-API meeting today and its a bit less clear as I may not have gotten a good test case. Will try to update the test case a bit later today.

**Comment by mhdawson** [comment]
Ok so the test case I had above was missing the right version (which I had checked as well) to replicate.  So the move complete test case is:

**Comment by mhdawson** [comment]
```JavaScript
const addon = require('../rwar')
const vm = require('vm');

try {
  addon.rwar();
} catch (err) {
  console.log('Regular context:' +  (err instanceof TypeError));
}

try {
  const contextObject = { };
  vm.runInNewContext("addon.rwar()", contextObject);
} catch (err) {
  console.log('New context:' + (err instanceof TypeError));
}

const contextObject = { console };
vm.runInNewContext('try {addon.rwar() } catch (err) {console.log("New context:" + (err instanceof TypeError))}', contextObject);
```

**Comment by mhdawson** [comment]
And more complete with an explanation for each. Still same conclusion that the native addon not being aware of the context when run under `runInNewContext` is related to the reported issue.

```JavaScript
const addon = require('../rwar')
const vm = require('vm');
const contextObject = { console };

/////////////////////////
// plain JavaScript
/////////////////////////
console.log('Plain JavaScript');

// expect err to be instnace of TypeError as only one context
try {
  throw new TypeError();
} catch (err) {
  console.log('Regular context:' +  (err instanceof TypeError));
}

// Don't expect err to be instance of TypeError as it was
// created in a different context from the context in which the
// check was made
try {
  const contextObject = { };
  vm.runInNewContext("throw new TypeError()", contextObject);
} catch (err) {
  console.log('New context:' + (err instanceof TypeError));
}

// expect err to be instance of TypeError as check is in
// same context as where err wsa created
vm.runInNewContext('try {throw new TypeError()} catch (err) {console.log("New context:" + (err instanceof TypeError))}', contextObject);

/////////////////////////
// Addons
/////////////////////////
console.log();
console.log('Addons');

// expect err to be instnace of TypeError as only one context
try {
  addon.rwar();
} catch (err) {
  console.log('Regular context:' +  (err instanceof TypeError));
}

// Don't expect err to be instance of TypeError as it was
// created in a different context from the context in which the
// check was made
try {
  const contextObject = { };
  vm.runInNewContext("addon.rwar()", contextObject);
} catch (err) {
  console.log('New context:' + (err instanceof TypeError));
}

// expect err to be instance of TypeError as check is in
// same context as where err wsa created
vm.runInNewContext('try {addon.rwar() } catch (err) {console.log("New context:" + (err instanceof TypeError))}', contextObject);
```

**Comment by mhdawson** [comment]
@gabrielschulhof updated test case as discussed in the N-API team meeting today.

From some additional experimental is seems like for addons the context used to create the error when a native method is run under  `vm.runInNewContext` is not the context passed in, nor the context outside of the `vm.runInNewContext` 

For example

```JavaScript

const copyOfTypeError = TypeError;
const contextObject = { console, TypeError, copyOfTypeError };
vm.runInNewContext('try {addon.rwar() } catch (err) {console.log("New context:" + (err instanceof copyOfTypeError))}', contextObject);
```

Still results in false, where as doing the same for plan JavaScript reports true.

**Comment by rynz** [comment]
Fantastic work @mhdawson!

**Comment by gabrielschulhof** [comment]
@mhdawson I also did some testing, and it looks like the addon always produces objects descended from the built-ins present in the context of its creation:

```C
#include <node_api.h>
#include "../../js-native-api/common.h"

static napi_value newError(napi_env env, napi_callback_info info) {
  size_t argc = 1;
  napi_value message, error;
  NAPI_CALL(env, napi_get_cb_info(env, info, &argc, &message, NULL, NULL));
  NAPI_CALL(env, napi_create_error(env, NULL, message, &error));
  return error;
}

NAPI_MODULE_INIT() {
  napi_property_descriptor methods[] = {
    DECLARE_NAPI_PROPERTY("newError", newError),
  };

  NAPI_CALL(env, napi_define_properties(
      env, exports, sizeof(methods) / sizeof(methods[0]), methods));

  return exports;
}
```

```JS
'use strict';

const binding = require(`./build/${common.buildType}/test_vm_context`);
const vm = require('vm');

const jsError = new Error('JS Error');
const nativeError = binding.newError('Native Error');

console.log('js:');
console.log('  created in: normal context, checked in: normal context: ' +
  (jsError instanceof Error));
console.log('  created in: normal context, checked in: vm context:     ' +
  vm.runInNewContext('jsError instanceof Error', { jsError }));
console.log('  created in: vm context,     checked in: normal context: ' +
  (vm.runInNewContext('new Error(\'JS Error\');') instanceof Error));
console.log('  created in: vm context,     checked in: vm context:     ' +
  vm.runInNewContext('(new Error(\'JS Error\')) instanceof Error;'));

console.log('');

console.log('native:');
console.log('  created in: normal context, checked in: normal context: ' +
  (nativeError instanceof Error));
console.log('  created in: normal context, checked in: vm context:     ' +
  vm.runInNewContext('nativeError instanceof Error', { nativeError }));
console.log('  created in: vm context,     checked in: normal context: ' +
  (vm.runInNewContext('binding.newError(\'nativeError\');', { binding })
    instanceof Error));
console.log('  created in: vm context,     checked in: vm context:     ' +
  vm.runInNewContext('binding.newError(\'nativeError\') instanceof Error;',
    { binding }));
```

and the output was:

```
js:
  created in: normal context, checked in: normal context: true
  created in: normal context, checked in: vm context:     false
  created in: vm context,     checked in: normal context: false
  created in: vm context,     checked in: vm context:     true

native:
  created in: normal context, checked in: normal context: true
  created in: normal context, checked in: vm context:     false
  created in: vm context,     checked in: normal context: true
  created in: vm context,     checked in: vm context:     false
```

Notice that I had to pass `binding` into the vm context. Ideally, I would `vm.runInNewContext('require("/path/to/binding").newError("Native Error");')`, meaning I would load a copy of the native add-on inside the vm context. `require()` is currently not defined in the vm context. I even tried passing the `require` into the vm context, but still, the add-on created instances descended from the built-ins of the outside context.

Basically, `NAPI_ADDON_INIT()` would have to be called again inside the vm context. This does not currently happen. I guess this is what @addaleax [meant](https://github.com/nodejs/node-addon-api/issues/764#issuecomment-658457203).

**Comment by gabrielschulhof** [comment]
I looked at the V8 documentation and, AFAICT it doesn't even have the facilities for creating a context-sensitive error:

https://v8docs.nodesource.com/node-14.1/da/d6a/classv8_1_1_exception.html

**Comment by mhdawson** [comment]
@gabrielschulhof 

>I looked at the V8 documentation and, AFAICT it doesn't even have the facilities for creating a context-sensitive error:

from the limited time I'd taken to look at it, I was thinking the same thing since the creation of TypeError had not way to pass in a context.

**Comment by mhdawson** [comment]
I did wonder though if there was some other way that the current context used by TypeError() could be set on the isolate or through some other way but had not had time to look at the code for that.

**Comment by addaleax** [comment]
@mhdawson @gabrielschulhof You could enter another context while creating the error object, if you want, but that’s probably not a good approach to take here, partially because N-API functions are more or less baked in stone and partially because this is not a typical use case (and because it would be better addressed through actually making N-API addons context-sensitive).

What you can do in this situation is passing the `TypeError` function from a context directly and creating an instance of that using `napi_new_instance`, that should not be an issue.

**Comment by mhdawson** [comment]
> and because it would be better addressed through actually making N-API addons context-sensitive
+1 

I think @gabrielschulhof and myself were just tying to think about how you might do that, as we don't have a good enough understanding of that yet.

**Comment by mhdawson** [comment]
From the discussion in https://github.com/nodejs/node-addon-api/issues/764 it sound like this is the expected behaviour for native addons when using vm.runInNewContext() and is not specific to node-addon-api. 

@rynz is ok if we close this out?

**Comment by mhdawson** [comment]
Going to close this out, please let us know if this is not the right thing to do.
